### PR TITLE
Unit Tests for JSON Serialization

### LINKED ISSUE #106: Thoroughly unit tests JSON Serialization
Check out TODO in `unit_test.py`.

### PR DESCRIPTION
**Pull request by aniketinfocusp**
Fixes #106. Unit tests for `to_json_serializable_type()` in `weather_mv/loader_pipeline/utils.py`. 

The current implementation of `to_json_serializable_type()`  throws error when a `ndarray` of size > 1 is passed.  Also, it did not handle arrays with NaNs and numbers mixed. 
Both of these issues are fixed in this PR.

### CODE REVIEW

File: weather_mv/loader_pipeline/util.py [hunk 1]
```diff
@@ -85,13 +85,15 @@ def to_json_serializable_type(value: t.Any) -> t.Any:
     # Note: The order of processing is significant.
     logger.debug('Serializing to JSON')
 
-    if pd.isna(value) or value is None:
+    # pd.isna() returns ndarray if input is not scalar therefore checking if not ndarray.
+    if type(value) != np.ndarray and (pd.isna(value) or value is None):
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/util.py:89]: What do you think about the following? 
```suggestion
    if (np.isscalar(value) and pd.isna(value)) or value is None:
```

If you prefer your method, my recommendation is to use `isinstance()` instead of type equality. I've been meaning to clean this up for a bit.

File: weather_mv/loader_pipeline/util.py [hunk 2]
```diff
@@ -85,13 +85,15 @@ def to_json_serializable_type(value: t.Any) -> t.Any:
...
         return None
     elif np.issubdtype(type(value), np.floating):
         return float(value)
     elif type(value) == np.ndarray:
         # Will return a scaler if array is of size 1, else will return a list.
-        return value.tolist()
+        # Replace all NaNs, NaTs with None.
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/util.py:95]: As mentioned above: can we replace all `type(x) == klass` with `isinstance(x, klass)`?

Comment (alxmrs) [inline weather_mv/loader_pipeline/util.py:95]: Nice! Good thinking.

File: weather_mv/loader_pipeline/util_test.py [hunk 3]
```diff
@@ -178,5 +185,99 @@ def test_make_attrs_ee_compatible_a2(self):
...
+
+    def test_to_json_serializable_type_none(self):
+        attrs = None
+
+        expected = None
+
+        actual = to_json_serializable_type(attrs)
+
+        self.assertEqual(actual, expected)
+        self.assertEqual(type(actual), type(expected))
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/util_test.py:197]: I don't think we need to include a type check; equality should be sufficient

File: weather_mv/loader_pipeline/util_test.py [hunk 4]
```diff
@@ -178,5 +185,99 @@ def test_make_attrs_ee_compatible_a2(self):
...
-    pass
+
+    def test_to_json_serializable_type_none(self):
+        attrs = None
+
+        expected = None
+
+        actual = to_json_serializable_type(attrs)
+
+        self.assertEqual(actual, expected)
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/util_test.py:196]: I know AAA is standard, but I think it would be more readable if we had a series of checks that all fit on one line that were mostly the same, where we only needed to check in the value being compared against. For example: 

```suggestion
    self.assertIsNone(self.convert(None))
    self.assertIsNone(self.convert(float('NaN')))
    self.assertIsNone(self.convert(np.nan))
    self.assertIsNotNone(self.convert([]))
```

File: weather_mv/loader_pipeline/util_test.py [hunk 5]
```diff
@@ -178,5 +185,99 @@ def test_make_attrs_ee_compatible_a2(self):
...
+        self.assertEqual(type(actual), type(expected))
+
+    def test_to_json_serializable_type_npfloat(self):
+        attrs = np.float32(1)
+
+        expected = float(1)
+
+        actual = to_json_serializable_type(attrs)
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/util_test.py:204]: Similar to my suggestion above. To help you out, here are a bunch of cases to check: 

```suggestion
    self.assertIsInstance(self.convert(0.1), float)
    self.assertIsInstance(self.convert(np.float32('0.1')), float)
    self.assertIsInstance(self.convert(np.float32('1')), float)
    self.assertIsInstance(self.convert(np.float16('0.1')), float)
    self.assertIsInstance(self.convert(np.single('0.1')), float)
    self.assertIsInstance(self.convert(np.double('0.1')), float)
    self.assertNotIsInstance(self.convert(1), float)
    self.assertNotIsInstance(self.convert(np.csingle('0.1')), float)
    self.assertNotIsInstance(self.convert(np.cdouble('0.1')), float)
    self.assertNotIsInstance(self.convert(np.intc('1')), float)
```

File: weather_mv/loader_pipeline/util_test.py [hunk 6]
```diff
@@ -178,5 +185,99 @@ def test_make_attrs_ee_compatible_a2(self):
...
+        self.assertEqual(type(actual), type(expected))
+
+    def test_to_json_serializable_type_ndarray(self):
+        attrs = np.arange(5)
+
+        expected = [0, 1, 2, 3, 4]
+
+        actual = to_json_serializable_type(attrs)
+
+        self.assertEqual(actual, expected)
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/util_test.py:216]: OK, I'll make an admission: I've written tests for these internally, but I have not yet had the time to port them to be external. I've not made these up on the fly, and I really appreciate that you're taking the initiative to fix this open issue :)

```suggestion
        self.assertIsInstance(self.convert(np.array(list(range(10)))), list)
    self.assertEqual(self.convert(np.array(list(range(10)))), list(range(10)))
    self.assertEqual(self.convert(np.array([1])), [1])
    self.assertEqual(
        self.convert(np.array([[1, 2, 3], [4, 5, 6]])), [[1, 2, 3], [4, 5, 6]]
    )
```

File: weather_mv/loader_pipeline/util_test.py [hunk 7]
```diff
@@ -178,5 +185,99 @@ def test_make_attrs_ee_compatible_a2(self):
...
+        self.assertEqual(type(actual), type(expected))
+
+    def test_to_json_serializable_type_ndarray_sizeone(self):
+        attrs = np.arange(1)
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/util_test.py:220]: This is a good idea for a test.

File: weather_mv/loader_pipeline/util_test.py [hunk 8]
```diff
@@ -178,5 +185,99 @@ def test_make_attrs_ee_compatible_a2(self):
...
+    def test_to_json_serializable_type_nptimedelta(self):
+        attrs = np.timedelta64(1, 'm')  # timedelta of 1 minute.
+
+        expected = float(60)  # seconds.
+
+        actual = to_json_serializable_type(attrs)
+
+        self.assertEqual(actual, expected)
+        self.assertEqual(type(actual), type(expected))
+
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/util_test.py:274]: Here are my internal tests that you're free to port here: 

```

  def test_time_to_isostring(self):
    now = datetime.datetime.now()
    now_str = now.isoformat()
    now_np = np.datetime64(now)

    self.assertEqual(self.convert(now), self.convert(now_str))
    self.assertEqual(self.convert(now), self.convert(now_np))

  def test_np_timedelta_to_float(self):
    delta = datetime.timedelta(seconds=33)
    delta_np = np.timedelta64(delta)
    self.assertEqual(self.convert(delta_np), delta.total_seconds())

  def test_np_integer_to_int(self):
    self.assertEqual(self.convert(np.int64(1)), 1)
    self.assertEqual(self.convert(np.int64(-10_000)), -10_000)
    self.assertEqual(self.convert(np.uint16(25)), 25)
```

File: weather_mv/loader_pipeline/util_test.py [hunk 9]
```diff
@@ -178,5 +185,99 @@ def test_make_attrs_ee_compatible_a2(self):
...
+    def test_to_json_serializable_type_npinteger(self):
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/util_test.py:275]: I like this idea for a test. Can you think of more corner cases to check?

**Review by alxmrs — COMMENTED** [review]
Added some feedback. Thanks for your fixes here!

File: weather_mv/loader_pipeline/util.py (hunk 1 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util.py:89]: I think using `np.isscalar` looks much cleaner and it handles other sequence objects like sets and tuples.

File: weather_mv/loader_pipeline/util_test.py (hunk 4 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util_test.py:196]: I agree, it looks much better and compressed when we have the checks inline. I went ahead and created a wrapper function `_convert` to wrap `to_json_serializable_type`.

File: weather_mv/loader_pipeline/util.py (hunk 1 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util.py]: Updated in https://github.com/google/weather-tools/pull/327/commits/dac23a047e2357b4f0dce9fb0b17969ae7f972a5.

File: weather_mv/loader_pipeline/util.py (hunk 2 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util.py:95]: Updated in https://github.com/google/weather-tools/pull/327/commits/dac23a047e2357b4f0dce9fb0b17969ae7f972a5.

File: weather_mv/loader_pipeline/util_test.py (hunk 3 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util_test.py]: Updated in https://github.com/google/weather-tools/pull/327/commits/d9cf97add436e462d01fb43fe2d32ac7c0511d17.

File: weather_mv/loader_pipeline/util_test.py (hunk 4 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util_test.py]: Added in https://github.com/google/weather-tools/pull/327/commits/d9cf97add436e462d01fb43fe2d32ac7c0511d17.

File: weather_mv/loader_pipeline/util_test.py (hunk 5 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util_test.py]: Thanks for these cases. Added in https://github.com/google/weather-tools/pull/327/commits/d9cf97add436e462d01fb43fe2d32ac7c0511d17.

File: weather_mv/loader_pipeline/util_test.py (hunk 6 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util_test.py]: I added some more I could think of. Added in https://github.com/google/weather-tools/pull/327/commits/d9cf97add436e462d01fb43fe2d32ac7c0511d17.

File: weather_mv/loader_pipeline/util_test.py (hunk 8 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util_test.py]: Thanks.

File: weather_mv/loader_pipeline/util_test.py (hunk 9 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util_test.py]: I have added some more cases (and used all that you have provided).

File: weather_mv/loader_pipeline/util.py [hunk 10]
```diff
@@ -85,14 +85,19 @@ def to_json_serializable_type(value: t.Any) -> t.Any:
     # Note: The order of processing is significant.
     logger.debug('Serializing to JSON')
 
-    if pd.isna(value) or value is None:
+    # pd.isna() returns ndarray if input is not scalar therefore checking if not ndarray.
```
Comment (alxmrs) [inline weather_mv/loader_pipeline/util.py:88]: Comment is a bit stale.

**Review by alxmrs — APPROVED** [review]
LGTM, thanks!

File: weather_mv/loader_pipeline/util.py (hunk 10 above)
Comment (aniketinfocusp) [reply weather_mv/loader_pipeline/util.py]: updated in https://github.com/google/weather-tools/pull/327/commits/af655e7d4c63b3dc2e39e87271f5528016981d50
